File: src/utils/proration.ts
import { Subscription, BillingCycle } from '../types/subscription';
import { InvoiceLineItem } from '../types/invoice';

export interface ProrationPreview {
  amount: number;
  isCredit: boolean;
  remainingDays: number;
  periodDays: number;
  oldDailyRate: number;
  newDailyRate: number;
  description: string;
  effectiveDate: 'immediate' | 'end_of_period';
}

export interface CreditMemo {
  subscriptionId: string;
  amount: number;
  reason: string;
  createdAt: Date;
  applied: boolean;
  remainingBalance: number;
}

const DAYS_IN_CYCLE: Record<BillingCycle, number> = {
  [BillingCycle.WEEKLY]: 7,
  [BillingCycle.MONTHLY]: 30,
  [BillingCycle.YEARLY]: 365,
  [BillingCycle.CUSTOM]: 30,
};

/**
 * Calculate days in a billing cycle
 */
export function getPeriodDays(cycle: BillingCycle): number {
  return DAYS_IN_CYCLE[cycle] ?? 30;
}

/**
 * Calculate remaining days in current billing period
 */
export function getRemainingDays(subscription: Subscription): number {
  const now = new Date();
  const nextBilling = new Date(subscription.nextBillingDate);
  const diffMs = nextBilling.getTime() - now.getTime();
  return Math.max(0, Math.ceil(diffMs / (1000 * 60 * 60 * 24)));
}

/**
 * Resolve the effective proration date for a plan change.
 *
 * If a specific date is provided, only immediate changes that happen before the
 * next billing date are prorated. Future-dated changes at or after the next bill
 * are treated as end-of-period changes.
 */
export function resolveProrationEffectiveDate(
  currentSubscription: Subscription,
  effectiveDate: 'immediate' | 'end_of_period' | Date = 'immediate'
): 'immediate' | 'end_of_period' {
  if (effectiveDate === 'end_of_period') {
    return 'end_of_period';
  }

  if (effectiveDate instanceof Date) {
    const nextBilling = new Date(currentSubscription.nextBillingDate);
    const now = new Date();
    if (
      effectiveDate.getTime() > now.getTime() &&
      effectiveDate.getTime() <= nextBilling.getTime()
    ) {
      return 'immediate';
    }
    return 'end_of_period';
  }

  return 'immediate';
}

/**
 * Calculate a prorated adjustment against the exact days remaining in the cycle.
 * This is the explicit mid-cycle engine used for plan upgrades and downgrades.
 */
export function calculateMidCycleProration(
  currentSubscription: Subscription,
  newPrice: number,
  effectiveDate: 'immediate' | 'end_of_period' | Date = 'immediate'
): ProrationPreview {
  const resolvedEffectiveDate = resolveProrationEffectiveDate(currentSubscription, effectiveDate);
  const periodDays = getPeriodDays(currentSubscription.billingCycle);

  if (resolvedEffectiveDate === 'end_of_period' || currentSubscription.price === newPrice) {
    return {
      amount: 0,
      isCredit: false,
      remainingDays: 0,
      periodDays,
      oldDailyRate: Math.round((currentSubscription.price / periodDays) * 100) / 100,
      newDailyRate: Math.round((newPrice / periodDays) * 100) / 100,
      description: 'No proration required',
      effectiveDate: 'end_of_period',
    };
  }

  const now = new Date();
  const nextBilling = new Date(currentSubscription.nextBillingDate);
  const chosenDate = effectiveDate instanceof Date ? effectiveDate : now;
  const targetDate = new Date(
    Math.min(Math.max(chosenDate.getTime(), now.getTime()), nextBilling.getTime())
  );
  const remainingMs = Math.max(0, nextBilling.getTime() - targetDate.getTime());
  const remainingDays = Math.max(0, Math.ceil(remainingMs / (1000 * 60 * 60 * 24)));

  const rawAmount = ((newPrice - currentSubscription.price) * remainingDays) / periodDays;
  const amount = Math.round(Math.abs(rawAmount) * 100) / 100;
  const isCredit = rawAmount < 0;

  const description =
    amount === 0
      ? 'No proration required'
      : isCredit
        ? `Prorated credit of ${amount} for plan downgrade (${remainingDays} days remaining)`
        : `Prorated charge of ${amount} for plan upgrade (${remainingDays} days remaining)`;

  return {
    amount,
    isCredit,
    remainingDays,
    periodDays,
    oldDailyRate: Math.round((currentSubscription.price / periodDays) * 100) / 100,
    newDailyRate: Math.round((newPrice / periodDays) * 100) / 100,
    description,
    effectiveDate: resolvedEffectiveDate,
  };
}

/**
 * Preview proration before confirming plan change
 *
 * Formula: (newRate - oldRate) * remainingDays / periodDays
 */
export function previewProration(
  currentSubscription: Subscription,
  newPrice: number,
  effectiveDate: 'immediate' | 'end_of_period' = 'immediate'
): ProrationPreview {
  return calculateMidCycleProration(currentSubscription, newPrice, effectiveDate);
}

/**
 * Calculate immediate upgrade with prorated charge
 */
export function calculateUpgradeProration(
  currentSubscription: Subscription,
  newPrice: number
): ProrationPreview {
  return previewProration(currentSubscription, newPrice, 'immediate');
}

/**
 * Calculate immediate downgrade with prorated credit
 */
export function calculateDowngradeProration(
  currentSubscription: Subscription,
  newPrice: number
): ProrationPreview {
  return previewProration(currentSubscription, newPrice, 'immediate');
}

/**
 * Calculate end-of-period change (no proration)
 */
export function calculateEndOfPeriodChange(
  currentSubscription: Subscription,
  newPrice: number
): ProrationPreview {
  return previewProration(currentSubscription, newPrice, 'end_of_period');
}

/**
 * Generate credit memo for downgrade
 */
export function generateCreditMemo(
  subscriptionId: string,
  amount: number,
  reason: string
): CreditMemo {
  return {
    subscriptionId,
    amount,
    reason,
    createdAt: new Date(),
    applied: false,
    remainingBalance: amount,
  };
}

/**
 * Apply credit memo to reduce charge amount
 */
export function applyCreditMemo(
  chargeAmount: number,
  creditMemo: CreditMemo
): { finalCharge: number; updatedMemo: CreditMemo } {
  if (creditMemo.applied || creditMemo.remainingBalance <= 0) {
    return { finalCharge: chargeAmount, updatedMemo: creditMemo };
  }

  const creditToApply = Math.min(chargeAmount, creditMemo.remainingBalance);
  const newRemaining = creditMemo.remainingBalance - creditToApply;

  return {
    finalCharge: Math.round((chargeAmount - creditToApply) * 100) / 100,
    updatedMemo: {
      ...creditMemo,
      remainingBalance: Math.round(newRemaining * 100) / 100,
      applied: newRemaining <= 0,
    },
  };
}

/**
 * Handle multiple changes within one cycle
 */
export function calculateNetProration(
  currentSubscription: Subscription,
  priceChanges: {
    oldPrice: number;
    newPrice: number;
    effectiveDate: 'immediate' | 'end_of_period';
  }[]
): ProrationPreview {
  let netAmount = 0;
  let remainingDays = getRemainingDays(currentSubscription);

  for (const change of priceChanges) {
    const result = calculateMidCycleProration(
      { ...currentSubscription, price: change.oldPrice },
      change.newPrice,
      change.effectiveDate
    );
    netAmount += result.isCredit ? -result.amount : result.amount;
    remainingDays = Math.max(remainingDays, result.remainingDays);
  }

  const isCredit = netAmount < 0;
  const amount = Math.round(Math.abs(netAmount) * 100) / 100;

  return {
    amount,
    isCredit,
    remainingDays,
    periodDays: getPeriodDays(currentSubscription.billingCycle),
    oldDailyRate: 0,
    newDailyRate: 0,
    description:
      amount === 0
        ? 'No net proration for multiple plan changes'
        : `Net ${isCredit ? 'credit' : 'charge'} of ${amount} for multiple plan changes`,
    effectiveDate: 'immediate',
  };
}

/**
 * Check if proration rounds to zero
 */
export function isZeroProration(preview: ProrationPreview): boolean {
  return preview.amount === 0;
}

/**
 * Build proration line item for invoice
 */
export function buildProrationLineItem(
  preview: ProrationPreview,
  currency: string
): InvoiceLineItem {
  return {
    description: preview.description,
    quantity: 1,
    unitPrice: preview.amount,
    currency,
    exchangeRate: 1,
    taxRateBps: 0, // Prorations typically not taxed separately
    lineTotal: preview.isCredit ? -preview.amount : preview.amount,
  };
}

/**
 * Proration configuration per plan
 */
export interface ProrationPlanConfig {
  planId: string;
  method: 'daily' | 'hourly' | 'none';
  allowMidCycleChanges: boolean;
  maxChangesPerCycle: number;
  creditExpirationDays: number;
  prorationPrecision: number;
}

export const DEFAULT_PRORATION_CONFIG: ProrationPlanConfig = {
  planId: 'default',
  method: 'daily',
  allowMidCycleChanges: true,
  maxChangesPerCycle: 3,
  creditExpirationDays: 90,
  prorationPrecision: 2,
};

/**
 * Proration analytics data
 */
export interface ProrationAnalyticsData {
  totalProrations: number;
  totalCreditsIssued: number;
  totalChargesApplied: number;
  netProrationAmount: number;
  averageProrationAmount: number;
  prorationCount: number;
  creditsUsedCount: number;
  chargesCount: number;
}

/**
 * Calculate proration with hourly granularity
 */
export function calculateHourlyProration(
  currentSubscription: Subscription,
  newPrice: number,
  effectiveDate: 'immediate' | 'end_of_period' = 'immediate'
): ProrationPreview {
  const periodDays = getPeriodDays(currentSubscription.billingCycle);
  const periodHours = periodDays * 24;
  const remainingDays =
    effectiveDate === 'end_of_period' ? 0 : getRemainingDays(currentSubscription);
  const remainingHours = remainingDays * 24;

  const rawAmount =
    effectiveDate === 'end_of_period'
      ? 0
      : ((newPrice - currentSubscription.price) * remainingHours) / periodHours;

  const amount = Math.round(Math.abs(rawAmount) * 100) / 100;
  const isCredit = rawAmount < 0;

  let description: string;
  if (amount === 0) {
    description = 'No proration required';
  } else if (isCredit) {
    description = `Prorated credit of ${amount} for plan downgrade (${remainingHours} hours remaining)`;
  } else {
    description = `Prorated charge of ${amount} for plan upgrade (${remainingHours} hours remaining)`;
  }

  return {
    amount,
    isCredit,
    remainingDays,
    periodDays,
    oldDailyRate: Math.round((currentSubscription.price / periodDays) * 100) / 100,
    newDailyRate: Math.round((newPrice / periodDays) * 100) / 100,
    description,
    effectiveDate,
  };
}

/**
 * Build proration analytics from a list of previews
 */
export function buildProrationAnalytics(previews: ProrationPreview[]): ProrationAnalyticsData {
  let totalCredits = 0;
  let totalCharges = 0;
  let creditsCount = 0;
  let chargesCount = 0;

  for (const preview of previews) {
    if (preview.isCredit) {
      totalCredits += preview.amount;
      creditsCount++;
    } else if (preview.amount > 0) {
      totalCharges += preview.amount;
      chargesCount++;
    }
  }

  const totalProrations = previews.length;
  const allAmounts = previews.map((p) => p.amount).filter((a) => a > 0);
  const averageAmount =
    allAmounts.length > 0
      ? Math.round((allAmounts.reduce((s, a) => s + a, 0) / allAmounts.length) * 100) / 100
      : 0;

  return {
    totalProrations,
    totalCreditsIssued: totalCredits,
    totalChargesApplied: totalCharges,
    netProrationAmount: Math.round((totalCharges - totalCredits) * 100) / 100,
    averageProrationAmount: averageAmount,
    prorationCount: totalProrations,
    creditsUsedCount: creditsCount,
    chargesCount,
  };
}

/**
 * Check if a subscription has exceeded its per-cycle change limit
 */
export function hasExceededChangeLimit(
  changeHistory: { applied: boolean; timestamp: Date }[],
  maxChangesPerCycle: number,
  cycleStart: Date
): boolean {
  const changesInCycle = changeHistory.filter(
    (c) => c.applied && c.timestamp.getTime() >= cycleStart.getTime()
  );
  return changesInCycle.length >= maxChangesPerCycle;
}

/**
 * Check if a credit memo has expired
 */
export function isCreditMemoExpired(creditMemo: CreditMemo, expirationDays: number): boolean {
  if (creditMemo.applied) return true;
  const now = new Date();
  const createdAt = new Date(creditMemo.createdAt);
  const diffMs = now.getTime() - createdAt.getTime();
  const diffDays = diffMs / (1000 * 60 * 60 * 24);
  return diffDays > expirationDays;
}

File: contracts/subscription/src/proration.rs
use soroban_sdk::{Address, Env, String};
use subtrackr_types::{Interval, Subscription};

/// Proration calculation result
#[soroban_sdk::contracttype]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ProrationResult {
    /// Amount to charge (positive) or credit (negative)
    pub amount: i128,
    /// Number of days remaining in current billing period
    pub remaining_days: u64,
    /// Total days in the billing period
    pub period_days: u64,
    /// The prorated daily rate of the old plan
    pub old_daily_rate: i128,
    /// The prorated daily rate of the new plan
    pub new_daily_rate: i128,
    /// Whether this is a charge or credit
    pub is_credit: bool,
    /// Human-readable description
    pub description: String,
}

/// Effective date for plan changes
#[derive(Clone, Debug, PartialEq)]
pub enum EffectiveDate {
    /// Change takes effect immediately
    Immediate,
    /// Change takes effect at the end of the current billing period
    EndOfPeriod,
}

/// Calculate the number of days in a billing interval
fn interval_days(interval: &Interval) -> u64 {
    match interval {
        Interval::Daily => 1,
        Interval::Weekly => 7,
        Interval::Monthly => 30,
        Interval::Quarterly => 90,
        Interval::Yearly => 365,
    }
}

/// Calculate proration for a plan change
/// 
/// Formula: (new_rate - old_rate) * remaining_days / period_days
/// 
/// # Arguments
/// * `env` — Soroban environment for timestamp access
/// * `subscription` — Current subscription state
/// * `old_price` — Current plan price
/// * `new_price` — New plan price
/// * `effective_date` — When the change takes effect
/// 
/// # Returns
/// ProrationResult with calculated amounts
pub fn calculate_proration(
    env: &Env,
    subscription: &Subscription,
    old_price: i128,
    new_price: i128,
    effective_date: EffectiveDate,
) -> ProrationResult {
    let now = env.ledger().timestamp();
    let period_seconds = subscription.next_charge_at - subscription.last_charged_at;
    let period_days = period_seconds / 86400;
    
    let remaining_seconds = if effective_date == EffectiveDate::EndOfPeriod {
        0 // No proration if effective at end of period
    } else {
        subscription.next_charge_at.saturating_sub(now)
    };
    let remaining_days = remaining_seconds / 86400;
    
    let old_daily_rate = old_price / period_days as i128;
    let new_daily_rate = new_price / period_days as i128;
    
    let amount = if effective_date == EffectiveDate::EndOfPeriod {
        0
    } else {
        (new_price - old_price) * remaining_days as i128 / period_days as i128
    };
    
    let is_credit = amount < 0;
    let abs_amount = amount.abs();
    
    let description = if is_credit {
        String::from_str(env, "Prorated credit for plan downgrade")
    } else if amount > 0 {
        String::from_str(env, "Prorated charge for plan upgrade")
    } else {
        String::from_str(env, "No proration required")
    };
    
    ProrationResult {
        amount: abs_amount,
        remaining_days,
        period_days,
        old_daily_rate,
        new_daily_rate,
        is_credit,
        description,
    }
}

/// Preview proration without applying changes
/// 
/// Returns the ProrationResult for display to the user before confirmation
pub fn preview_proration(
    env: &Env,
    subscription: &Subscription,
    old_price: i128,
    new_price: i128,
    effective_date: EffectiveDate,
) -> ProrationResult {
    calculate_proration(env, subscription, old_price, new_price, effective_date)
}

/// Generate a credit memo for downgrade credits
/// 
/// Credit memos are stored on-chain and can be applied to future invoices
pub fn generate_credit_memo(
    env: &Env,
    subscription_id: u64,
    amount: i128,
    reason: String,
) -> CreditMemo {
    CreditMemo {
        subscription_id,
        amount,
        reason,
        created_at: env.ledger().timestamp(),
        applied: false,
    }
}

/// Credit memo structure for on-chain storage
#[soroban_sdk::contracttype]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CreditMemo {
    pub subscription_id: u64,
    pub amount: i128,
    pub reason: String,
    pub created_at: u64,
    pub applied: bool,
}

/// Apply a credit memo to reduce a charge amount
/// 
/// Returns the remaining charge after credit application
pub fn apply_credit_memo(
    charge_amount: i128,
    credit_memo: &mut CreditMemo,
) -> i128 {
    if credit_memo.applied || credit_memo.amount <= 0 {
        return charge_amount;
    }
    
    let credit_to_apply = charge_amount.min(credit_memo.amount);
    credit_memo.amount -= credit_to_apply;
    credit_memo.applied = credit_memo.amount == 0;
    
    charge_amount - credit_to_apply
}

/// Handle edge case: multiple changes within one cycle
/// 
/// When a user changes plans multiple times in one billing period,
/// we track the net proration across all changes
pub fn calculate_net_proration(
    env: &Env,
    subscription: &Subscription,
    price_changes: &[(i128, i128, EffectiveDate)], // (old_price, new_price, effective_date)
) -> ProrationResult {
    let mut total_amount: i128 = 0;
    
    for (old_price, new_price, effective_date) in price_changes {
        let result = calculate_proration(env, subscription, *old_price, *new_price, effective_date.clone());
        if result.is_credit {
            total_amount -= result.amount;
        } else {
            total_amount += result.amount;
        }
    }
    
    let is_credit = total_amount < 0;
    let abs_amount = total_amount.abs();
    
    let description = if is_credit {
        String::from_str(env, "Net prorated credit for multiple plan changes")
    } else if total_amount > 0 {
        String::from_str(env, "Net prorated charge for multiple plan changes")
    } else {
        String::from_str(env, "No net proration for plan changes")
    };
    
    ProrationResult {
        amount: abs_amount,
        remaining_days: 0, // Aggregate doesn't have a single remaining period
        period_days: interval_days(&Interval::Monthly), // Default
        old_daily_rate: 0,
        new_daily_rate: 0,
        is_credit,
        description,
    }
}

/// Handle zero-dollar prorations
/// 
/// Returns true if the proration rounds to zero
pub fn is_zero_proration(result: &ProrationResult) -> bool {
    result.amount == 0
}

/// Rounding accuracy helper
/// 
/// Ensures consistent rounding across the system
pub fn round_proration_amount(amount: i128, decimals: u32) -> i128 {
    let factor = 10i128.pow(decimals);
    (amount + factor / 2) / factor * factor
}